File: src/maze/algos/random_two/Interface.py
import random


class Interface(object):

    def __init__(self, maze):
        self.__maze = maze
        self.__success = True

    @property
    def width(self):
        return len(self.__maze)
    
    @property
    def height(self):
        return len(self.__maze[0])

    @property
    def success(self):
        return self.__success

    def get_random_float(self):
        return random.random()

    def set_wall(self, x, y, direction, wall_exists):

        # Sanity checks
        if not 0 <= x < self.width:
            print("Error: x value {} is invalid".format(x))
            self.__success = False
            return
        if not 0 <= y < self.height:
            print("Error: y value {} is invalid".format(y))
            self.__success = False
            return
        if len(direction) != 1 or direction not in 'nesw':
            print("Error: direction value '{}' is invalid".format(direction))
            self.__success = False
            return

        # Set the wall value
        self.__maze[x][y][direction] = wall_exists

        # Set the opposing wall value
        if direction == 'n' and y < self.height - 1:
            self.__maze[x][y + 1]['s'] = wall_exists
        if direction == 'e' and x < self.width - 1:
            self.__maze[x + 1][y]['w'] = wall_exists
        if direction == 's' and 0 < y:
            self.__maze[x][y - 1]['n'] = wall_exists
        if direction == 'w' and 0 < x:
            self.__maze[x - 1][y]['e'] = wall_exists

File: src/maze/templates/java/Interface.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

public class Interface {

    ArrayList<ArrayList<HashMap<Character, Boolean>>> _maze;
    Random _random;
    BooleanHolder _success;

    public Interface(
        ArrayList<ArrayList<HashMap<Character, Boolean>>> maze,
        Random random,
        BooleanHolder success
    ) {
        _maze = maze;
        _random = random;
        _success = success;
    }

    int getWidth() {
        return _maze.size();
    }

    int getHeight() {
        return _maze.get(0).size();
    }

    double getRandomFloat() {
        return _random.nextDouble();
    }

    void setWall(int x, int y, char direction, boolean wallExists) {

        // Sanity checks
        if (x < 0 || getWidth() <= x) {
            System.out.println("Error: x value " + x + " is invalid");
            _success.set(false);
            return;
        }
        if (y < 0 || getHeight() <= y) {
            System.out.println("Error: x value " + y + " is invalid");
            _success.set(false);
            return;
        }
        if (direction != 'n' && direction != 'e' && direction != 's' && direction != 'w') {
            System.out.println(
                "Error: direction value '" + direction + "' is invalid"
            );
            _success.set(false);
            return;
        }

        // Set the wall value
        _maze.get(x).get(y).put(direction, wallExists);

        // Set the opposing wall value
        switch (direction) {
            case 'n': {
                if (y < getHeight() - 1) {
                    _maze.get(x).get(y + 1).put('s', wallExists);
                }
                break;
            }
            case 'e': {
                if (x < getWidth() - 1) {
                    _maze.get(x + 1).get(y).put('w', wallExists);
                }
                break;
            }
            case 's': {
                if (0 < y) {
                    _maze.get(x).get(y - 1).put('n', wallExists);
                }
                break;
            }
            case 'w': {
                if (0 < x) {
                    _maze.get(x - 1).get(y).put('e', wallExists);
                }
                break;
            }
        }
    }
}
